### PR TITLE
Directmapping

### PR DESCRIPTION
**Pull request by ihartb**
- inverted PT, 
- direct allocation and deallocation
- cannot allocate more than a single page on a give myallocate call (as this would require swapping, so stack for threads uses OG malloc and free as of now)
- returns `NULL` to processes other than library that have allocated more than a page

### CODE REVIEW

File: asst2/CODE/datastructs.c [hunk 1]
```diff
@@ -157,14 +158,14 @@ void free_map(hashmap* h) {
...
       // context, linkedlists free'd earlier
-      free(node->value); //free tcb
-      free(node);
+      mydeallocate(node->value, __FILE__, __LINE__, LIBRARYREQ); //free tcb
+      free(node); //calloc'd
       node = next;
     }
   }
   
   free(h->buckets);
```
Comment (Victor-C-Zhang) [inline asst2/CODE/datastructs.c]: is this an oops?

File: asst2/CODE/my_malloc.c [hunk 2]
```diff
@@ -4,39 +4,251 @@
...
+		write_free_page(pdata, i);
+		metadata* mdata = (metadata*) ((char*)mem_space + i*page_size);
+		write_free_size(mdata, num_segments);
+	}
+}
+
+
+//returns to the user a pointer to the amount of memory requested
+void* myallocate(size_t size, char* file, int line, int threadreq){
+  	enter_scheduler(&timer_pause_dump);
```
Comment (Victor-C-Zhang) [inline asst2/CODE/my_malloc.c]: does this work if we don't have a scheduler inited?

File: asst2/CODE/my_malloc.h [hunk 3]
```diff
@@ -2,8 +2,32 @@
...
+#include "my_scheduler.h" 
 
-void* myallocate(size_t size);
-void mydeallocate(void* ptr);
+#define THREADREQ 1
+#define LIBRARYREQ 0
+//#define malloc(x) myallocate(x, __FILE__, __LINE__, THREADREQ)
+//#define free(x) mydeallocate(x, __FILE__, __LINE__, THREADREQ)
+#define MEMSIZE 8388608
+#define SEGMENTSIZE 64
```
Comment (Victor-C-Zhang) [inline asst2/CODE/my_malloc.h]: Can you pull in the global variables for malloc here? They're in `global_vals.h`

File: asst2/CODE/datastructs.c (hunk 1 above)
Comment (ihartb) [reply asst2/CODE/datastructs.c]: nah, two different values being free'd, one uses calloc and one uses myallocate

File: asst2/CODE/my_malloc.c (hunk 2 above)
Comment (ihartb) [reply asst2/CODE/my_malloc.c]: o lel hmm hasn't given an issue actually idk

File: asst2/CODE/my_malloc.h (hunk 3 above)
Comment (ihartb) [reply asst2/CODE/my_malloc.h]: hmm wait, so move `MEMSIZE` and `SEGMENTSIZE` into global_vals.h? or move the globals from my_malloc.c into my_malloc.h?

Comment (Victor-C-Zhang) [reply asst2/CODE/my_malloc.h]: Yes, can you put these as const values in `global_vals`?

File: asst2/CODE/datastructs.c (hunk 1 above)
Comment (Victor-C-Zhang) [reply asst2/CODE/datastructs.c]: I don't think we can use calloc either

File: asst2/CODE/my_malloc.c [hunk 4]
```diff
@@ -4,39 +4,251 @@
...
+//and return this so is_occupied = false
+//else we will return 0x80 which is true 
+int is_occupied(metadata* curr) {
+	return  (curr->size) & 0x80;
+}
+
+//ignores the leftmost bit in the metadata which represents free/malloc
+//to return the number of userdata blocks
+unsigned char block_size(metadata* curr) {
+	return curr -> size & 0x7f;
```
Comment (Victor-C-Zhang) [inline asst2/CODE/my_malloc.c:35]: this is incorrect. 0x0 represents 1, not 0. We need to add 1, otherwise everything breaks.

Comment (Victor-C-Zhang) [reply asst2/CODE/my_malloc.c:35]: unless you've already factored this in elsewhere, in which case we'll be fine

Comment (Victor-C-Zhang) [reply asst2/CODE/my_malloc.c:35]: actually, I'm tripping. ignore this whole thread ^
I need sleep

Comment (Victor-C-Zhang) [reply asst2/CODE/my_malloc.c:35]: although we may want to revisit this in the future. If we can allocate more than a page at a time, we should have a flag that tells us that the allocation is more than a page. Then we'll have only 6 bits per metadata to store segment length. Then we'd have to do this normalization technique.

Comment (ihartb) [reply asst2/CODE/my_malloc.c:35]: yeah, there might have to be several metadata / pagedata changes to account for multiple page allocations, might have to discuss em

**Review by Victor-C-Zhang — COMMENTED** [review]
Good to go now, only need to merge in from main

**Pull request merged**